R/convert_to_date.R

```diff
@@ -0,0 +1,151 @@
+                                               date_system="modern",
+                                               include_time=NULL,
+                                               round_seconds=TRUE,
+                                               tz="UTC",
+                                               out_class=c("POSIXct", "Date")) {
+  if (!is.null(include_time)) {
+    warning("`include_time` is ignored in favor of `out_class`.")
+  }
+  out_class <- match.arg(out_class)
+  excel_numeric_to_date(
+    date_num=x,
+    date_system="modern",
+    round_seconds=round_seconds,
+    tz=tz,
+    include_time=out_class %in% "POSIXct"
+  )
+}
+
+convert_to_datetime_helper.factor <- function(x, ..., out_class=c("POSIXct", "Date")) {
+  convert_to_datetime_helper.character(as.character(x), ..., out_class=out_class)
+}
+
+convert_to_datetime_helper.POSIXt <- function(x, ..., out_class=c("POSIXct", "Date")) {
+  out_class <- match.arg(out_class)
+  if (out_class %in% "POSIXct") {
+    # Ensure that POSIXlt gets converted to POSIXct
+    as.POSIXct(x, ...)
+  } else {
+    as.Date(x, ...)
+  }
+}
+
+convert_to_datetime_helper.Date <- function(x, ..., tz="UTC", out_class=c("POSIXct", "Date")) {
+  out_class <- match.arg(out_class)
+  if (out_class %in% "POSIXct") {
+    ret <- as.POSIXct(x, ...)
+    # as.POSIXct.Date ignores the time zone, so manually apply it.
+    attr(ret, "tzone") <- tz
+  } else {
+    ret <- x
+  }
+  ret
+}
+
+convert_to_datetime_helper.character <- function(x, ..., tz="UTC", character_fun=lubridate::ymd_hms, allow_na_conversion=FALSE, out_class=c("POSIXct", "Date")) {
+  out_class <- match.arg(out_class)
+  mask_na <- is.na(x)
+  mask_excel_numeric <- !mask_na & grepl(pattern="^[0-9]{5}(?:\\.[0-9]*)?$", x=x)
+  mask_character <- !(mask_na | mask_excel_numeric)
+  if (out_class %in% "POSIXct") {
+    ret <- as.POSIXct(x=rep(NA, length(x)), tz="UTC")
+  } else {
+    ret <- as.Date(x=rep(NA, length(x)))
+  }
+  if (any(mask_excel_numeric)) {
+    ret[mask_excel_numeric] <- convert_to_datetime_helper(as.numeric(x[mask_excel_numeric]), ..., tz=tz)
+  }
+  if (any(mask_character)) {
+    characters_converted <-
+      if (out_class %in% "POSIXct") {
+        character_fun(x[mask_character], tz=tz, ...)
+      } else {
+        character_fun(x[mask_character], ...)
+      }
+    if (!(out_class %in% class(characters_converted))) {
+      stop(
+        "`character_fun(x)` must return class ", out_class,
+        "; the returned class was: ", paste(class(characters_converted), collapse=", ")
+      )
+    }
+    ret[mask_character] <- characters_converted
+    if (any(is.na(ret[mask_character]))) {
+      not_converted_values <- unique(x[mask_character & is.na(ret)])
+      not_converted_message <-
+        paste0(
+          "Not all character strings converted to class ", out_class,
+          ".  Values not converted were: ",
+          paste0('"', not_converted_values, '"', collapse=", ")
```

Comment:
Is it possible/likely that this could be say, a thousand values, if this function were called incorrectly and would that cause a crash?  Maybe truncate at some point - there's code in `get_dupes` for that?

Reply:
Good call.  I added that.